cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -397,46 +399,48 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
    */
   public List<MetadataEntry> search(final String namespaceId, final String searchQuery,
                                     final MetadataSearchTargetType type) {
-    Set<MetadataEntry> results = new HashSet<>();
+    List<MetadataEntry> results = new ArrayList<>();
     Scanner scanner;
 
-    String namespacedSearchQuery = prepareSearchQuery(namespaceId, searchQuery);
-
-    if (namespacedSearchQuery.endsWith("*")) {
-      // if prefixed search get start and stop key
-      byte[] startKey = Bytes.toBytes(namespacedSearchQuery.substring(0, namespacedSearchQuery.lastIndexOf("*")));
-      byte[] stopKey = Bytes.stopKeyForPrefix(startKey);
-      scanner = indexedTable.scanByIndex(Bytes.toBytes(INDEX_COLUMN), startKey, stopKey);
-    } else {
-      byte[] value = Bytes.toBytes(namespacedSearchQuery);
-      scanner = indexedTable.readByIndex(Bytes.toBytes(INDEX_COLUMN), value);
-    }
-    try {
-      Row next;
-      while ((next = scanner.next()) != null) {
-        String rowValue = next.getString(INDEX_COLUMN);
-        if (rowValue == null) {
-          continue;
-        }
+    for (String queryWord : Splitter.on(SPACE_SEPARATOR_PATTERN).omitEmptyStrings().trimResults().split(searchQuery)) {
+      String namespacedSearchQuery = prepareSearchQuery(namespaceId, queryWord);
 
-        final byte[] rowKey = next.getRow();
-        String targetType = MdsKey.getTargetType(rowKey);
-
-        // Filter on target type if not ALL
-        if ((type != MetadataSearchTargetType.ALL) && (type !=
-          MetadataSearchTargetType.valueOfSerializedForm(targetType))) {
-          continue;
+      if (namespacedSearchQuery.endsWith("*")) {
```

Comment:
so this will now support queries of the form ``data* with  field     name*`` as well?